File: ios-client/Sources/Features/TeslaAPI.swift
import Foundation

struct VehicleState: Codable {
    let batteryLevel: Int
    let chargingState: String
    let latitude: Double
    let longitude: Double
    let speed: Double?
    let timestamp: Date
    let range: Double
}

enum TeslaError: Error {
    case authenticationFailed
    case networkError
    case invalidResponse
}

class TeslaAPI {
    private let backendBaseURL: String
    private var accessToken: String?
    private var selectedVehicleId: String?

    init(backendURL: String = "https://api.tesla.com") {
        self.backendBaseURL = backendURL
        // Try to load saved token
        self.accessToken = UserDefaults.standard.string(forKey: "teslaAccessToken")
    }

    func authenticate(username: String, password: String) async throws -> Bool {
        let url = URL(string: "\(backendBaseURL)/oauth/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "grant_type": "password",
            "client_id": "ownerapi",
            "email": username,
            "password": password,
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200,
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let token = json["access_token"] as? String
        else {
            throw TeslaError.authenticationFailed
        }

        self.accessToken = token
        UserDefaults.standard.set(token, forKey: "teslaAccessToken")
        return true
    }

    func getVehicles() async throws -> [[String: Any]] {
        guard let token = accessToken else {
            throw TeslaError.authenticationFailed
        }

        let url = URL(string: "\(backendBaseURL)/api/1/vehicles")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200,
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let vehicles = json["response"] as? [[String: Any]]
        else {
            throw TeslaError.invalidResponse
        }

        return vehicles
    }

    func getVehicleState() async throws -> VehicleState {
        guard let token = accessToken, let vehicleId = selectedVehicleId else {
            throw TeslaError.authenticationFailed
        }

        let url = URL(string: "\(backendBaseURL)/api/1/vehicles/\(vehicleId)/vehicle_data")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200,
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let response = json["response"] as? [String: Any],
            let driveState = response["drive_state"] as? [String: Any],
            let chargeState = response["charge_state"] as? [String: Any]
        else {
            throw TeslaError.invalidResponse
        }

        return VehicleState(
            batteryLevel: chargeState["battery_level"] as? Int ?? 0,
            chargingState: chargeState["charging_state"] as? String ?? "Disconnected",
            latitude: driveState["latitude"] as? Double ?? 0,
            longitude: driveState["longitude"] as? Double ?? 0,
            speed: driveState["speed"] as? Double,
            timestamp: Date(),
            range: chargeState["battery_range"] as? Double ?? 0
        )
    }

    func setVehicle(id: String) {
        self.selectedVehicleId = id
    }

    func startCharging() async throws {
        guard let vehicleId = selectedVehicleId else {
            throw TeslaError.invalidResponse
        }
        try await sendCommand("charge_start", vehicleId: vehicleId)
    }

    func stopCharging() async throws {
        guard let vehicleId = selectedVehicleId else {
            throw TeslaError.invalidResponse
        }
        try await sendCommand("charge_stop", vehicleId: vehicleId)
    }

    private func sendCommand(_ command: String, vehicleId: String) async throws {
        guard let token = accessToken else {
            throw TeslaError.authenticationFailed
        }

        let url = URL(string: "\(backendBaseURL)/api/1/vehicles/\(vehicleId)/command/\(command)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200
        else {
            throw TeslaError.networkError
        }
    }
}

File: awhittlewandering/consolidated/ContinentalUSA-mobile/utils/TeslaAPI.js
import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';

const BASE_URL = 'https://owner-api.teslamotors.com/api/1';

export class TeslaAPI {
  static async authenticate(email, password) {
    try {
      const response = await axios.post(`${BASE_URL}/oauth/token`, {
        grant_type: 'password',
        client_id: 'ownerapi',
        client_secret: 'c7257eb71a564034f9419ee651c7d0e5f7aa6bfbd18bafb5c5c033b093bb2fa3',
        email,
        password
      });
      
      await AsyncStorage.setItem('tesla_token', response.data.access_token);
      return response.data;
    } catch (error) {
      console.error('Tesla authentication error:', error);
      throw error;
    }
  }

  static async getVehicles() {
    const token = await AsyncStorage.getItem('tesla_token');
    try {
      const response = await axios.get(`${BASE_URL}/vehicles`, {
        headers: { 'Authorization': `Bearer ${token}` }
      });
      return response.data.response;
    } catch (error) {
      console.error('Error fetching vehicles:', error);
      throw error;
    }
  }

  static async getVehicleData(vehicleId) {
    const token = await AsyncStorage.getItem('tesla_token');
    try {
      const response = await axios.get(`${BASE_URL}/vehicles/${vehicleId}/vehicle_data`, {
        headers: { 'Authorization': `Bearer ${token}` }
      });
      return response.data.response;
    } catch (error) {
      console.error('Error fetching vehicle data:', error);
      throw error;
    }
  }

  static async setTemperature(vehicleId, temp) {
    const token = await AsyncStorage.getItem('tesla_token');
    try {
      const response = await axios.post(
        `${BASE_URL}/vehicles/${vehicleId}/command/set_temps`,
        { driver_temp: temp },
        { headers: { 'Authorization': `Bearer ${token}` } }
      );
      return response.data.response;
    } catch (error) {
      console.error('Error setting temperature:', error);
      throw error;
    }
  }

  static async startHVAC(vehicleId) {
    const token = await AsyncStorage.getItem('tesla_token');
    try {
      const response = await axios.post(
        `${BASE_URL}/vehicles/${vehicleId}/command/auto_conditioning_start`,
        {},
        { headers: { 'Authorization': `Bearer ${token}` } }
      );
      return response.data.response;
    } catch (error) {
      console.error('Error starting HVAC:', error);
      throw error;
    }
  }
}
